### PR TITLE
Whitelist ott for graphql schema

### PR DESCRIPTION
**Pull request by alextargov**
**Description**

Changes proposed in this pull request:
- Expose the OTT for application when querying applications
- Do not return the one-time token for application when it is used or it is expired. This way, when someone queries the application and gets the system auths he wont see the OTT if expired/used. This will force him to generate a new valid one.

**Related issue(s)**

- ...

**Pull Request status**

- [x] Implementation
- [x] Unit tests
- [x] Integration tests
- [x] `chart/compass/values.yaml` is updated 
- [x] Mocks are regenerated, using the automated script

### CODE REVIEW

File: components/director/internal/domain/application/resolver.go [hunk 1]
```diff
@@ -501,22 +526,51 @@ func (r *Resolver) Auths(ctx context.Context, obj *graphql.Application) ([]*grap
...
 
 	err = tx.Commit()
 	if err != nil {
 		return nil, err
 	}
-
 	return out, nil
 }
 
+func (r *Resolver) checkApplicationOneTimeTokenIsInvalid(auth *model.Auth, config onetimetoken.Config) bool {
```
Comment (pankrator) [inline components/director/internal/domain/application/resolver.go:557]: maybe `isApplicationOTTInvalid`

File: components/director/internal/domain/application/resolver.go [hunk 2]
```diff
@@ -501,22 +526,51 @@ func (r *Resolver) Auths(ctx context.Context, obj *graphql.Application) ([]*grap
...
+			log.C(ctx).Debug("skipping one-time token due to its expiration or usage")
+			continue
+		}
+
 		c, err := r.sysAuthConv.ToGraphQL(&sa)
 		if err != nil {
 			return nil, err
 		}
 
+		if sa.Value.OneTimeToken != nil && sa.Value.OneTimeToken.Type == tokens.ApplicationToken {
```
[the remark this answers is not in the document]
Comment (alextargov) [reply components/director/internal/domain/application/resolver.go:539]: Not sure if we'd like to expose the OTT for other types than Applications at the moment. Probably if there's a need to expose them as well, it won't be much of an effort.

File: components/director/internal/domain/application/resolver.go [hunk 3]
```diff
@@ -501,22 +526,51 @@ func (r *Resolver) Auths(ctx context.Context, obj *graphql.Application) ([]*grap
...
+	}
+
+	if auth.OneTimeToken.Used {
+		return true
+	}
+
+	isExpired := auth.OneTimeToken.CreatedAt.Add(config.ApplicationExpiration).Before(r.timeService.Now())
+	if isExpired {
+		return true
+	}
```
Comment (pankrator) [inline components/director/internal/domain/application/resolver.go:569]: There is the following portion of code in file `validation_hydrator.go`:
```
expirationTime, err := vh.getExpirationTimeForToken(systemAuth)
	if err != nil {
		log.C(ctx).Error(err)
		respondWithAuthSession(ctx, w, authSession)
		return
	}

	if systemAuth.Value.OneTimeToken.CreatedAt.Add(expirationTime).Before(vh.timeService.Now()) {
		log.C(ctx).Infof("One Time Token with validity %s for system auth with ID %q has expired", expirationTime.String(), systemAuth.ID)
		respondWithAuthSession(ctx, w, authSession)
		return
	}
```

The code for expiration can be extracted in common function

File: components/director/pkg/graphql/schema.graphql [hunk 4]
```diff
@@ -742,6 +742,7 @@ type Auth {
 	additionalQueryParams: QueryParams
 	additionalQueryParamsSerialized: QueryParamsSerialized
 	requestAuth: CredentialRequestAuth
+	oneTimeToken: OneTimeTokenForApplication
```
Comment (pankrator) [inline components/director/pkg/graphql/schema.graphql:745]: In my opinion, this should work for all kind of OTT, not only Application, as the type is Auth, which is generic.

File: tests/connector/tests/hydrator_api_test.go [hunk 5]
```diff
@@ -93,13 +96,33 @@ func TestHydrators(t *testing.T) {
...
+			}
 		})
 
 		t.Run("should resolve certificate for "+testCase.clientType, func(t *testing.T) {
 			//given
-			token, err := testCase.tokenGenerationFunc(t, testCase.clientId)
+			var err error
+			var token externalschema.Token
+
+			app, err := fixtures.RegisterApplicationFromInput(t, ctx, directorClient.DexGraphqlClient, cfg.Tenant, graphql.ApplicationRegisterInput{
```
Comment (pankrator) [inline tests/connector/tests/hydrator_api_test.go:113]: Why was it needed to move app creation for each test case and check inside the clientID which has to be passed to `tokenGenerationFunc`?

**Review by pankrator — CHANGES_REQUESTED** [review]

File: tests/connector/tests/hydrator_api_test.go [hunk 6]
```diff
@@ -23,33 +23,22 @@ func TestHydrators(t *testing.T) {
 	require.NotEmpty(t, runtime.ID)
 	runtimeID := runtime.ID
 
-	app, err := fixtures.RegisterApplicationFromInput(t, ctx, directorClient.DexGraphqlClient, cfg.Tenant, graphql.ApplicationRegisterInput{
```
Comment (ataleksandrov) [inline tests/connector/tests/hydrator_api_test.go:26]: Not sure, why this is needed?

File: components/director/pkg/graphql/schema.graphql (hunk 4 above)
[the remark this answers is not in the document]
Comment (alextargov) [reply components/director/pkg/graphql/schema.graphql:745]: No, that would be generic. You would do 

` oneTimeToken {
            ... on OneTimeTokenForApplication {
              token
              connectorURL
              legacyConnectorURL
            }
          }`

or 
` oneTimeToken {
            ... on OneTimeTokenForRuntime {
              token
              connectorURL
            }
          }`

File: components/director/internal/domain/onetimetoken/service.go [hunk 7]
```diff
@@ -297,3 +303,54 @@ func (s *service) getSuggestedTokenForApp(ctx context.Context, app *model.Applic
...
+		return time.Duration(0), errors.Errorf("One Time Token for system auth id %s has no valid type", systemAuth.ID)
+	}
+}
+
+func (s *service) IsTokenValid(systemAuth *model.SystemAuth) (bool, error) {
+	if systemAuth.Value == nil {
+		return false, errors.Errorf("System Auth value for auth id %s is missing", systemAuth.ID)
+	}
+
+	if !s.isAuthOneTimeTokenType(systemAuth) {
```
Comment (ataleksandrov) [inline components/director/internal/domain/onetimetoken/service.go:325]: Could we just check `systemAuth.Value.OneTimeToken == nil`?

File: components/director/internal/oathkeeper/validation_hydrator.go [hunk 8]
```diff
@@ -3,49 +3,44 @@ package oathkeeper
...
 // ValidationHydrator missing godoc
 type ValidationHydrator interface {
 	ResolveConnectorTokenHeader(w http.ResponseWriter, r *http.Request)
 }
 
-// Service missing godoc
-//go:generate mockery --name=Service --output=automock --outpkg=automock --case=underscore
-type Service interface {
+// SystemAuthService missing godoc
+//go:generate mockery --name=SystemAuthService --output=automock --outpkg=automock --case=underscore
```
Comment (ataleksandrov) [inline components/director/internal/oathkeeper/validation_hydrator.go:22]: Could you please delete the previous mock file?

File: tests/connector/tests/hydrator_api_test.go [hunk 9]
```diff
@@ -31,25 +23,30 @@ func TestHydrators(t *testing.T) {
...
+	defer fixtures.CleanupRuntime(t, ctx, directorClient.DexGraphqlClient, cfg.Tenant, &runtime)
+	require.NoError(t, err)
+	require.NotEmpty(t, runtime.ID)
+	runtimeID := runtime.ID
+
 	hash := "df6ab69b34100a1808ddc6211010fa289518f14606d0c8eaa03a0f53ecba578a"
 
 	for _, testCase := range []struct {
 		clientType           string
-		clientId             string
```
Comment (ataleksandrov) [inline tests/connector/tests/hydrator_api_test.go:38]: Having unified `cliet_id`, removes the need of` if testCase.clientType == "Application" {` everywhere except in the check here: https://github.com/kyma-incubator/compass/pull/2007/files#diff-c5b72ac3ff812b64b53e2ea366f61ea7068307ff8b5e7e8ce07b5e19e4de84daR104 , so I would propose to return it.

File: components/director/internal/api/token_resolver.go [hunk 10]
```diff
@@ -21,6 +21,7 @@ type TokenResolver interface {
 //go:generate mockery --name=TokenService --output=automock --outpkg=automock --case=underscore
 type TokenService interface {
 	RegenerateOneTimeToken(ctx context.Context, authID string, token tokens.TokenType) (model.OneTimeToken, error)
+	IsTokenValid(systemAuth *model.SystemAuth) (bool, error)
```
Comment (pankrator) [inline components/director/internal/api/token_resolver.go:24]: This seems like it is not used

File: components/director/internal/oathkeeper/validation_hydrator_test.go [hunk 11]
```diff
@@ -272,13 +203,14 @@ func TestValidationHydrator_ResolveConnectorTokenHeader(t *testing.T) {
...
-		validationHydrator := NewValidationHydrator(tokenService, transact, timeService, csrTokenExpiration, appTokenExpiration, runtimeTokenExpiration)
+		validationHydrator := NewValidationHydrator(tokenService, transact, oneTimeTokenService)
 		authenticationSession := connector.AuthenticationSession{}
 		req := createAuthRequestWithTokenQueryParam(t, authenticationSession, token)
 		w := httptest.NewRecorder()
+		oneTimeTokenService.On("IsTokenValid", systemAuth).Return(true, nil)
 		tokenService.On("GetByToken", mock.Anything, token).Return(systemAuth, nil)
 		tokenService.On("InvalidateToken", mock.Anything, mock.Anything).Return(nil)
-		timeService.On("Now").Return(beforeOneDay)
+		//timeService.On("Now").Return(beforeOneDay)
```
Comment (pankrator) [inline components/director/internal/oathkeeper/validation_hydrator_test.go:213]: Delete this

File: components/director/internal/oathkeeper/validation_hydrator_test.go [hunk 12]
```diff
@@ -302,14 +233,15 @@ func TestValidationHydrator_ResolveConnectorTokenHeader(t *testing.T) {
...
+		validationHydrator := NewValidationHydrator(tokenService, transact, oneTimeTokenService)
 
 		authenticationSession := connector.AuthenticationSession{}
 		req := createAuthRequestWithTokenQueryParam(t, authenticationSession, token)
 		w := httptest.NewRecorder()
+		oneTimeTokenService.On("IsTokenValid", systemAuth).Return(true, nil)
 		tokenService.On("GetByToken", mock.Anything, token).Return(systemAuth, nil)
 		tokenService.On("InvalidateToken", mock.Anything, mock.Anything).Return(nil)
-		timeService.On("Now").Return(beforeOneDay)
+		//timeService.On("Now").Return(beforeOneDay)
```
Comment (pankrator) [inline components/director/internal/oathkeeper/validation_hydrator_test.go:244]: Leftover?

File: tests/connector/tests/hydrator_api_test.go [hunk 13]
```diff
@@ -93,13 +98,26 @@ func TestHydrators(t *testing.T) {
...
+			}
 		})
 
 		t.Run("should resolve certificate for "+testCase.clientType, func(t *testing.T) {
 			//given
-			token, err := testCase.tokenGenerationFunc(t, testCase.clientId)
+			var err error
+			var token externalschema.Token
+
+			if testCase.clientType == "Application" {
```
Comment (pankrator) [inline tests/connector/tests/hydrator_api_test.go:115]: As Acho pointed out this snippets can be simplified (like it was before) as they are identical in each test case

**Review by pankrator — CHANGES_REQUESTED** [review]

**Review by pankrator — APPROVED** [review]

**Pull request merged**